src/klein/test/test_headers.py

```diff
@@ -0,0 +1,548 @@
+        # type: (bytes) -> None
+        """
+        L{headerValueAsText} decodes L{bytes} using L{HEADER_VALUE_ENCODING}.
+        """
+        self.assertEqual(headerValueAsText(value), decodeValue(value))
+
+
+    @given(text(min_size=1))
+    def test_headerValueAsTextWithText(self, value):
+        # type: (Text) -> None
+        """
+        L{headerValueAsText} passes through L{Text}.
+        """
+        self.assertIdentical(headerValueAsText(value), value)
+
+
+
+class HeaderNameNormalizationTests(TestCase):
+    """
+    Tests for header name normalization.
+    """
+
+    def test_normalizeLowerCase(self):
+        # type: () -> None
+        """
+        L{normalizeHeaderName} normalizes header names to lower case.
+        """
+        self.assertEqual(normalizeHeaderName("FooBar"), "foobar")
+
+
+
+class RawHeadersConversionTests(TestCase):
+    """
+    Tests for L{normalizeRawHeaders}.
+    """
+
+    def test_pairNotTuple(self):
+        # type: () -> None
+        """
+        L{normalizeRawHeaders} raises L{TypeError} if the C{headerPairs}
+        argument is not an tuple L{tuple}s.
+        """
+        e = self.assertRaises(
+            TypeError, tuple, normalizeRawHeaders(([b"k", b"v"],))
+        )
+        self.assertEqual(str(e), "header pair must be a tuple")
+
+
+    def test_pairsWrongLength(self):
+        # type: () -> None
+        """
+        L{normalizeRawHeaders} raises L{ValueError} if the C{headerPairs}
+        argument is not an tuple of 2-item L{tuple}s.
+        """
+        for pair in ((b"k",), (b"k", b"v", b"x")):
+            e = self.assertRaises(
+                ValueError,
+                tuple,
+                normalizeRawHeaders(cast(Iterable[Iterable[bytes]], (pair,))),
+            )
+            self.assertEqual(str(e), "header pair must be a 2-tuple")
+
+
+    def test_pairsNameNotBytes(self):
+        # type: () -> None
+        """
+        L{normalizeRawHeaders} raises L{TypeError} if the C{headerPairs}
+        argument is not an tuple of 2-item L{tuple}s where the first item in
+        the 2-item L{tuple} is L{bytes}.
+        """
+        e = self.assertRaises(
+            TypeError,
+            tuple,
+            normalizeRawHeaders(
+                cast(Iterable[Iterable[bytes]], ((u"k", b"v"),))
+            )
+        )
+        self.assertEqual(str(e), "header name must be bytes")
+
+
+    def test_pairsValueNotBytes(self):
+        # type: () -> None
+        """
+        L{normalizeRawHeaders} raises L{TypeError} if the C{headerPairs}
+        argument is not an tuple of 2-item L{tuple}s where the second item
+        in the 2-item L{tuple} is bytes.
+        """
+        e = self.assertRaises(
+            TypeError,
+            tuple,
+            normalizeRawHeaders(
+                cast(Iterable[Iterable[bytes]], ((b"k", u"v"),))
+            )
+        )
+        self.assertEqual(str(e), "header value must be bytes")
+
+
+
+class GetValuesTestsMixIn(object):
```

Comment:
I'm also moving in the inheritance suck model, but `TestCase` isn't really designed to support that well, and I'd like to avoid having to write a bunch of stubby test methods than call into common code or figure out some other way to inject tests when this makes it easy, and since it's purely internal (in tests, even) and not a public interface, it seems reasonable without a known simple alternative.

I'm open to suggestions here, but absent that, don't have a better plan.